### PR TITLE
plot 5th, 50th and 95th percentile of value and derivate instead of p…

### PR DESCRIPTION
**Pull request by LvHang**
@danpovey 
Hi Dan,
1.I print the percentiles instead of the mean+stddev in the plots.
2.I have already tested it using 1-4 exp-dirs respectively. I'm sure the lenged will not cover the figures. The result in /export/a11/hlyu/kaldi/egs/swbd/s5c/report{1..4} respectively. You can check them.
3.In code, I still reserved the mean/stddev values and just commented the code about ploting them. Because I think it will be easy for others use it if they want.

Best wishes,
Hang

### CODE REVIEW

**Comment by danpovey** [comment]
thanks.
best to delete the old code, people can get it from the git history if they want.

**Comment by LvHang** [comment]
Ok, another question about the format of nonlinearity log file in report directory.
The original format is 'Iteration\tValueMean\tValueStddev\tDerivMean\tDerivStddev\n'
You hope the new format is 'Iteration\tValue_5th\tValue_50th\tValue_95th\tDeriv_5th\tDeriv_50th\tDeriv_95th\n"'
or
'Iteration\tValueMean\tValueStddev\tDerivMean\tDerivStddev\tValue_5th\tValue_50th\tValue_95th\tDeriv_5th\tDeriv_50th\tDeriv_95th\n"'

**Comment by danpovey** [comment]
I don't think anyone will look at that file directly so it probably doesn't
matter.
Let me know when you think the PR is ready to merge.

**Comment by LvHang** [comment]
Hi Dan,
I have already deleted the code about ploting mean+-stddev.
And I choose the format "Iteration	ValueMean ValueStddev	DerivMean DerivStddev	Value_5th Value_50th Value_95th Deriv_5th Deriv_50th Deriv_95th" for log files.
I test them, the results are stored in /export/a11/hlyu/kaldi/egs/swbd/s5c/report{1..4} respectively.
Please check it. (I will send a example with 4 curves by email, I think it's easy for you to see.)

Best wishes,
Hang

**Comment by danpovey** [comment]
@LvHang, is this ready?

**Comment by LvHang** [comment]
@danpovey 
Hi Dan,
I update the code. Although the appearance looks ok, the code look duty. Let me consider how to change.
Bests,
Hang

**Comment by LvHang** [comment]
@danpovey 
Hi Dan,
Could you give me some suggestions about how to modifying the codes? Now I think the figure(output of the codes) could satisfy your requirement, but I think the code looks not satisfying.
Hang

**Comment by vijayaditya** [comment]
Hi Hang,
I had a quick look at the code. First impression is that you should reduce
the size of your functions. You could use the Google python style guide for
more guidance.

--Vijay

**Comment by LvHang** [comment]
@vijayaditya 
Thank you for your suggestion. Let me fix it.

**Comment by LvHang** [comment]
@danpovey 
Hi Dan,
As vija's suggestion--reduce the size of your functions. I modified the style of the python scripts. Move the shared codes to a standalone function as much as possible.
At the same time, I do some tests on lstm/non-lstm model. You can check the final result in /export/a11/hlyu/kaldi/egs/swbd/s5c/reprot[34][_non_lstm] directories.
Thanks a lot.
Hang

**Comment by danpovey** [comment]
Thanks!
I skimmed it very quickly and didn't see a problem.. if you say it's working I am inclined to believe it.  @vijayaditya, unless you have any immediate objection I'd merge this.

File: egs/wsj/s5/steps/libs/nnet3/report/log_parse.py [hunk 1]
```diff
@@ -51,36 +97,45 @@ def parse_progress_logs_for_nonlinearity_stats(exp_dir):
...
-            stats_per_component_per_iter[component_name] = {}
-            stats_per_component_per_iter[component_name][
-                'type'] = component_type
-            stats_per_component_per_iter[component_name]['stats'] = {}
-            stats_per_component_per_iter[component_name][
-                'stats'][iteration] = [value_mean, value_stddev,
-                                       deriv_mean, deriv_stddev]
-
+        if component_type == 'LstmNonlinearity':
+            parse_regex_lstmp = re.compile(
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/libs/nnet3/report/log_parse.py]: I think this function is still big. I recommend moving this code to a new function.

File: egs/wsj/s5/steps/nnet3/report/generate_plots.py [hunk 2]
```diff
@@ -279,64 +350,50 @@ def generate_nonlin_stats_plots(exp_dir, output_dir, plot, comparison_dir=None,
...
-                             bbox_to_anchor=(0.5, -0.5 + len(dirs) * -0.2),
-                             ncol=1, borderaxespad=0.)
-            plt.grid(True)
-            fig.suptitle("Mean and stddev of the value and derivative at "
-                         "{comp_name}".format(comp_name=component_name))
-            comp_name = latex_compliant_name(component_name)
-            figfile_name = '{dir}/nonlinstats_{comp_name}.pdf'.format(
-                dir=output_dir, comp_name=comp_name)
-            fig.savefig(figfile_name, bbox_extra_artists=(lgd,),
+            if stats_per_dir[exp_dir][component_name]['type'] == 'LstmNonlinearity':
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/nnet3/report/generate_plots.py]: This code could benefit from few comments. Sorry that I didn't add any even in the original code. Could you please do it in this commit ?

File: egs/wsj/s5/steps/libs/nnet3/report/log_parse.py [hunk 3]
```diff
@@ -27,6 +28,51 @@ def __init__(self, message = None):
                            "There was an error while trying to parse the logs."
                            " Details : \n{0}\n".format(message))
 
+# This function is used to fill stats_per_component_per_iter table with the 
+# results of regular expression.
+def fill_stats_table_with_regex_result(groups, gate_index, stats_table):
+    iteration = int(groups[0])
+    component_name = groups[1]
+    component_type = groups[2]
+    value_percentiles = groups[3+gate_index*6]
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/libs/nnet3/report/log_parse.py]: Do you think you could reduce this verbosity using a for loop and a dictionary ?

File: egs/wsj/s5/steps/libs/nnet3/report/log_parse.py [hunk 4]
```diff
@@ -27,6 +28,51 @@ def __init__(self, message = None):
                            "There was an error while trying to parse the logs."
                            " Details : \n{0}\n".format(message))
 
+# This function is used to fill stats_per_component_per_iter table with the 
+# results of regular expression.
+def fill_stats_table_with_regex_result(groups, gate_index, stats_table):
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/libs/nnet3/report/log_parse.py]: better rename the function to specify that these are non-linearity stats.

**Comment by LvHang** [comment]
@vijayaditya 
Thank you for your suggestions. Please check it.
(1) I think the main point of "parse_progress_logs_for_nonlinearity_stats" function looks still big is the regular expression is too long. So I move it out of the function.
(2) I summarized the steps of "generate_nonlin_stats_plots" function and add few comments at the beginning of the function.
(3) I rename the "fill_tats_table_with_regex_result" to "fill_nonlin_stats_table_with_regex_result" so that it could show the meaning of nonlinearity statistics.
(4) As I want to save the semantic of the variables in "fill_nonlin_stats_table_with_regex_result", so I didn't convert it to a loop and dict.

Best wishes,
Hang

**Comment by vijayaditya** [comment]
@danpovey LGTM

**Comment by danpovey** [comment]
Thanks!  Merging.

**Pull request merged**